extensions/s3/src/main/java/io/deephaven/extensions/s3/S3SeekableByteChannel.java

```diff
@@ -0,0 +1,321 @@
+/**
+ * Copyright (c) 2016-2022 Deephaven Data Labs and Patent Pending
+ */
+package io.deephaven.extensions.s3;
+
+import io.deephaven.UncheckedDeephavenException;
+import io.deephaven.base.verify.Assert;
+import java.util.concurrent.CancellationException;
+import io.deephaven.parquet.base.util.SeekableChannelsProvider;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import software.amazon.awssdk.services.s3.S3AsyncClient;
+import software.amazon.awssdk.services.s3.S3Uri;
+import software.amazon.awssdk.services.s3.model.HeadObjectResponse;
+
+import java.io.IOException;
+import java.net.URI;
+import java.nio.ByteBuffer;
+import java.nio.channels.ClosedChannelException;
+import java.nio.channels.NonWritableChannelException;
+import java.nio.channels.SeekableByteChannel;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+
+/**
+ * {@link SeekableByteChannel} class used to fetch objects from AWS S3 buckets using an async client with the ability to
+ * read ahead and cache fragments of the object.
+ */
+final class S3SeekableByteChannel implements SeekableByteChannel, SeekableChannelsProvider.ContextHolder {
+
+    private static final long CLOSED_SENTINEL = -1;
+
+    /**
+     * Context object used to store read-ahead buffers for efficiently reading from S3.
+     */
+    static final class S3ChannelContext implements SeekableChannelsProvider.ChannelContext {
+
+        /**
+         * Used to store information related to a single fragment
+         */
+        static class FragmentContext {
+            /**
+             * The index of the fragment in the object
+             */
+            private int fragmentIndex;
+
+            /**
+             * The future that will be completed with the fragment's bytes
+             */
+            private CompletableFuture<ByteBuffer> future;
+
+            private FragmentContext(final int fragmentIndex, final CompletableFuture<ByteBuffer> future) {
+                this.fragmentIndex = fragmentIndex;
+                this.future = future;
+            }
+        }
+
+        /**
+         * Used to cache recently fetched fragments for faster lookup
+         */
+        private final List<FragmentContext> bufferCache;
+
+        private long size;
+
+        S3ChannelContext(final int maxCacheSize) {
+            bufferCache = new ArrayList<>(maxCacheSize);
+            for (int i = 0; i < maxCacheSize; i++) {
+                bufferCache.add(null);
+            }
+            size = -1;
```

Comment:
I wanted to avoid growing the size at the time of sending requests and to be able to directly do cache[idx%size]. That is why I allocated everything upfront.  I have changed to array.